plugins/processors/aws/ec2/ec2.go

```diff
@@ -0,0 +1,210 @@
+
+	Log        telegraf.Logger     `toml:"-"`
+	tags       map[string]struct{} `toml:"-"`
+	imdsClient *imds.Client
+	parallel   parallel.Parallel
+}
+
+const sampleConfig = `
+  ## Tags to attach to metrics. Available tags:
+  ## * accountId
+  ## * architecture
+  ## * availabilityZone
+  ## * billingProducts
+  ## * imageId
+  ## * instanceId
+  ## * instanceType
+  ## * kernelId
+  ## * pendingTime
+  ## * privateIp
+  ## * ramdiskId
+  ## * region
+  ## * version
+  tags = []
+
+  ## Timeout for http requests made by against aws ec2 metadata endpoint.
+  timeout = "10s"
+
+  ## ordered controls whether or not the metrics need to stay in the same order
+  ## this plugin received them in. If false, this plugin will change the order
+  ## with requests hitting cached results moving through immediately and not
+  ## waiting on slower lookups. This may cause issues for you if you are
+  ## depending on the order of metrics staying the same. If so, set this to true.
+  ## Keeping the metrics ordered may be slightly slower.
+  ordered = false
+`
+
+const (
+	DefaultTimeout             = 10 * time.Second
+	DefaultMaxOrderedQueueSize = 10_000
+)
+
+var allowedTags = map[string]struct{}{
+	"accountId":        {},
+	"architecture":     {},
+	"availabilityZone": {},
+	"billingProducts":  {},
+	"imageId":          {},
+	"instanceId":       {},
+	"instanceType":     {},
+	"kernelId":         {},
+	"pendingTime":      {},
+	"privateIp":        {},
+	"ramdiskId":        {},
+	"region":           {},
+	"version":          {},
+}
+
+func (r *AwsEc2Processor) SampleConfig() string {
+	return sampleConfig
+}
+
+func (r *AwsEc2Processor) Description() string {
+	return "Attach AWS EC2 metadata to metrics"
+}
+
+func (r *AwsEc2Processor) Add(metric telegraf.Metric, acc telegraf.Accumulator) error {
+	r.parallel.Enqueue(metric)
+	return nil
+}
+
+func (r *AwsEc2Processor) Start(acc telegraf.Accumulator) error {
+	r.Log.Debug("Initializing AWS EC2 Processor")
+
+	if r.Timeout == 0 {
+		r.Timeout = config.Duration(DefaultTimeout)
+	}
+	if r.MaxParallelCalls == 0 {
+		r.MaxParallelCalls = 10_000
+	}
+
+	r.tags = make(map[string]struct{})
+	for _, tag := range r.Tags {
+		if len(tag) > 0 && isTagAllowed(tag) {
+			r.tags[tag] = struct{}{}
+		} else {
+			return fmt.Errorf(
+				"Not allowed metadata tag specified in configuration: %s", tag,
+			)
+		}
+	}
+
+	if len(r.tags) == 0 {
+		return errors.New("No allowed metadata tags specified in configuration")
+	}
+
+	if r.Ordered {
+		r.parallel = parallel.NewOrdered(acc, r.asyncAdd, DefaultMaxOrderedQueueSize, r.MaxParallelCalls)
+	} else {
+		r.parallel = parallel.NewUnordered(acc, r.asyncAdd, r.MaxParallelCalls)
+	}
+
+	cfg, err := awsconfig.LoadDefaultConfig(context.Background())
+	if err != nil {
+		return err
+	}
+
+	r.imdsClient = imds.NewFromConfig(cfg)
+
+	return nil
+}
+
+func (r *AwsEc2Processor) Stop() error {
+	if r.parallel == nil {
+		return errors.New("Trying to stop unstarted AWS EC2 Processor")
+	}
```

Comment:
Since `parallel` needs `acc` for construction I actually need to put this in `Start` (unless we want `Init()` to have a signature which would accept `acc telegraf.Accumulator`).